### ISSUE TITLE
pkg_install: pkg_remove before installing package can be dangerous

### ISSUE THREAD

**Issue by konimex** [kissx/kiss#8]
If a package that contains `/usr/bin/tar` is removed (in the "default" case, [busybox](https://github.com/kissx/packages/blob/ed2e07219bbd0d6336bf6118f30481ed163b5e37/repo/busybox/manifest#L401)), then `kiss` won't be able to proceed to the next step as `tar` doesn't even exist.

**Comment by vypr** [comment]
A theoretical alternative is to make `kiss` a package (which would make it easier to update btw) and require a dependency on something that provides `tar`. Isolating the dependency to busybox will not necessarily be good for the end-user who may want GNU's coreutils or suckless' tools.

busybox or another coreutils package could be provided in an install CD or something of the sort so that you could bootstrap <insert coreutils here> and `kiss` on a system. The only downside is that the coreutil package wouldn't be able to be uninstalled through kiss, it'd have to be manually replaced by the user, limiting ease of customization.

We don't want to necessarily bundle an individual `tar` with `kiss` itself, unless we could find some alternative compression/decompression that operates through `sh`.

**Comment by konimex** [comment]
>Isolating the dependency to busybox will not necessarily be good for the end-user who may want GNU's coreutils or suckless' tools.

I don't see how this is relevant to the problem. I only said default case since that's what the packages repo provides by default. If one wants to use GNU coreutils then as Dylan said in the KISS website, "as a user you can easily create your own repository and package what you like". Besides, all the tools KISS rely on are POSIX, and coreutils doesn't even provide `tar`.

**Comment by dylanaraps** [comment]
One solution is to unpack the archive somewhere first but then we'd still need `cp` to be available after `pkg_remove`. How do other package managers handle this?

Another solution is to unpack the `tar` (overwriting existing files) and afterwards remove anything not in the new manifest file (by comparing it to the old manifest file).

**Comment by konimex** [comment]
For reference, this is Slackware's [`installpkg`](http://termbin.com/vrbb) (from pkgtools) and Sabotage Linux's [`butch-core`](https://github.com/sabotage-linux/sabotage/blob/e3fd95cb1e94331a16cd0c7f4ce127b495e824b2/KEEP/bin/butch-core), both are written in POSIX sh.

**Comment by dylanaraps** [comment]
I'll take a look thanks. (Check your email for an organisation invite FYI). I'll play around with it locally and come up with a solution.

**Comment by dylanaraps** [comment]
Does `tar` even have a POSIX specification? 

I've come up with a rather simple solution: 

- In `pkg_remove` skip deleting any files called `tar`.
- Unpack the package tarball to `$sys_root/` and overwrite any (remaining) existing files.

This will ensure `tar` is available to unpack the tarball and will also ensure that the new version of `tar` is installed afterwards. There's a minor issue however. It seems to fail as `/bin/tar` is a symlink to `/bin/busybox` and `tar` fails to replace the symlink.

**Comment by dylanaraps** [comment]
I wonder if we should add a special case and statically compile a second `busybox` instance. One that won't be added to `$PATH` to remove this issue completely. This way all dependencies (minus `git` and `file`) will always be available to `kiss`.

**Comment by dylanaraps** [comment]
As this issue only affects `tar`, another solution is to copy the `tar` binary somewhere temporary (can limit it only to packages containing `tar`). This seems like the simplest to me.

**Comment by konimex** [comment]
>Does `tar` even have a POSIX specification?

As in the file format? Yes. As in the command? Not to my knowledge.

Say, apart from the obvious issue of overwriting `/etc` files (if they do exist in the package), what are the potential problems of overwriting the package?

Speaking of affecting, when it comes to `pkg_remove`, I think we need to be careful of both `rm` and `rmdir` too. Same problem with `tar` (Solely in regards to `pkg_install`, if someone wants to remove (not upgrade) then it's their problem).

(Also, I haven't received the e-mail, can you re-send it to me? Thanks.)

**Comment by dylanaraps** [comment]
> Say, apart from the obvious issue of overwriting /etc files (if they do exist in the package), what are the potential problems of overwriting the package?

It turns out that `tar` by default overwrites "old" files (we can use `-k` to disable this).

> (Also, I haven't received the e-mail, can you re-send it to me? Thanks.)

I sent the invite again. Check all emails on your GitHub account.

Thoughts on the above issue (I edited it to include example code).

**Comment by dylanaraps** [comment]
Seeing as the `tar` command has no POSIX specification I'm leaning more towards a second, static `busybox` instance.

**Comment by dylanaraps** [comment]
I'd like to avoid `pkg_remove` too if possible.

Here's what currently happens if you remove it.

```
./var/
./var/db/
./var/db/kiss/
./var/db/kiss/zlib/
./var/db/kiss/zlib/sources
tar: ./var/db/kiss/zlib/sources: Cannot open: File exists
./var/db/kiss/zlib/manifest
tar: ./var/db/kiss/zlib/manifest: Cannot open: File exists
./var/db/kiss/zlib/checksums
tar: ./var/db/kiss/zlib/checksums: Cannot open: File exists
./var/db/kiss/zlib/build
tar: ./var/db/kiss/zlib/build: Cannot open: File exists
./var/db/kiss/zlib/version
tar: ./var/db/kiss/zlib/version: Cannot open: File exists
tar: Exiting with failure status due to previous errors
=> Installed zlib#1.2.11-1
```

And.

```
./var/
./var/db/
./var/db/kiss/
./var/db/kiss/zlib/
./var/db/kiss/zlib/sources
tar: ./var/db/kiss/zlib/sources: Cannot open: File exists
./var/db/kiss/zlib/manifest
tar: ./var/db/kiss/zlib/manifest: Cannot open: File exists
./var/db/kiss/zlib/checksums
tar: ./var/db/kiss/zlib/checksums: Cannot open: File exists
./var/db/kiss/zlib/build
tar: ./var/db/kiss/zlib/build: Cannot open: File exists
./var/db/kiss/zlib/version
tar: ./var/db/kiss/zlib/version: Cannot open: File exists
tar: Exiting with failure status due to previous errors
=> Installed zlib#1.2.11-1
```

What's your email? I'm pretty sure I can send it directly.

**Comment by konimex** [comment]
Ah that's not good. So pretty much the second option, then.

>What's your email? I'm pretty sure I can send it directly.

herdiansyah [at] netc [dot] eu

**Comment by dylanaraps** [comment]
> Ah that's not good. So pretty much the second option, then.

I fixed it. I had included `-k` near the end of the command and didn't notice.

**Comment by dylanaraps** [comment]
Removing `-k` and overwriting all files will also remove everything in `/etc` though. Will open a PR with the code.

**Comment by konimex** [comment]
To tackle this, maybe we can move the affected `/etc` file temporarily (e.g. `mv file file.backup` or something like that during the untar stage and move it back after). Or maybe we can do the `pacman` way and install the new `/etc` file with `.kissnew` suffix, but that may require a temporary "unpackaged" directory.

**Comment by dylanaraps** [comment]
See #9.

**Comment by dylanaraps** [comment]
Fixed in #9.
